### DISCUSSION TITLE
Introspecting an existing database doesn't handle join tables correctly

### DISCUSSION THREAD

**Discussion by MorayM** [prisma/prisma#25067] [Q&A]
### Question

I'm in the process of adding Prisma 5 to an existing project with a large existing database (created with Prisma 1). I've introspected the database and edited the schema so that the names of the relation fields are cased correctly. I then created an initial migration SQL file as per the guide and marked it as applied.
When I try to create a new migration, I keep getting errors telling me that ALL of my existing JOIN tables are incorrect:
```
Drift detected: Your database schema is not in sync with your migration history.

The following is a summary of the differences between the expected database schema given your migrations files, and the actual schema of the database.

It should be understood as the set of changes to get from the expected schema to the actual schema.

[*] Changed the `_RoleToTenantUser` table
  [-] Removed foreign key on columns (A)
  [-] Removed foreign key on columns (B)
  [+] Added foreign key on columns (A)
  [+] Added foreign key on columns (B)
  [*] Renamed index `_RoleToTenantUser_AB_unique` to `RoleToTenantUser_AB_unique`
  [*] Renamed index `_RoleToTenantUser_B_index` to `B`

... continued for the other 87 join tables in my schema ...
```

What's going on here, and how do I fix it?

My (VERY reduced) generated Prisma schema is below:

```prisma
model Role {
  id                              String                @id @db.Char(25) @default(cuid())
  name                            String                @db.MediumText
  description                     String?               @db.MediumText
  createdAt                       DateTime
  updatedAt                       DateTime              @updatedAt()
  tenantUsers                     TenantUser[]

  @@index([owner], map: "owner")
}

model TenantUser {
  id                                String                @id @db.Char(25) @default(cuid())
  email                             String?               @db.MediumText
  roles                             Role[]
}
```
My existing table looks like this:

```sql
CREATE TABLE `_RoleToTenantUser` (
  `A` char(25) CHARACTER SET utf8 NOT NULL,
  `B` char(25) CHARACTER SET utf8 NOT NULL,
  UNIQUE KEY `RoleToTenantUser_AB_unique` (`A`,`B`),
  KEY `B` (`B`),
  CONSTRAINT `_RoleToTenantUser_ibfk_1` FOREIGN KEY (`A`) REFERENCES `Role` (`id`) ON DELETE CASCADE,
  CONSTRAINT `_RoleToTenantUser_ibfk_2` FOREIGN KEY (`B`) REFERENCES `TenantUser` (`id`) ON DELETE CASCADE
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci
```

And the SQL generated with `prisma migrate diff` for this relation is

```sql
CREATE TABLE `_RoleToTenantUser` (
    `A` CHAR(25) NOT NULL,
    `B` CHAR(25) NOT NULL,

    UNIQUE INDEX `_RoleToTenantUser_AB_unique`(`A`, `B`),
    INDEX `_RoleToTenantUser_B_index`(`B`)
) DEFAULT CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci;
ALTER TABLE `_RoleToTenantUser` ADD CONSTRAINT `_RoleToTenantUser_A_fkey` FOREIGN KEY (`A`) REFERENCES `Role`(`id`) ON DELETE CASCADE ON UPDATE CASCADE;
ALTER TABLE `_RoleToTenantUser` ADD CONSTRAINT `_RoleToTenantUser_B_fkey` FOREIGN KEY (`B`) REFERENCES `TenantUser`(`id`) ON DELETE CASCADE ON UPDATE CASCADE;
```

### Information about Prisma Schema, Client Queries and Environment (optional)

- OS: EndeavourOS
- Database: MySQL 5.7
- Node.js version: 16.20.2

- Run `prisma -v` to see your Prisma version and paste it
```
prisma                  : 5.18.0
@prisma/client          : 5.18.0
Computed binaryTarget   : debian-openssl-1.1.x
Operating System        : linux
Architecture            : x64
Node.js                 : v16.20.2
Query Engine (Node-API) : libquery-engine 4c784e32044a8a016d99474bd02a3b6123742169 (at node_modules/@prisma/engines/libquery_engine-debian-openssl-1.1.x.so.node)
Schema Engine           : schema-engine-cli 4c784e32044a8a016d99474bd02a3b6123742169 (at node_modules/@prisma/engines/schema-engine-debian-openssl-1.1.x)
Schema Wasm             : @prisma/prisma-schema-wasm 5.18.0-25.4c784e32044a8a016d99474bd02a3b6123742169
Default Engines Hash    : 4c784e32044a8a016d99474bd02a3b6123742169
Studio                  : 0.502.0
```

**Comment by ludralph** [comment]
Hi @MorayM 👋 

Thank you for raising this question.

It is good that you want to upgrade your Prisma 1 project. I'll recommend that you do read the [Prisma 1 upgrade guide](https://www.prisma.io/docs/orm/more/upgrade-guides/upgrade-from-prisma-1/upgrading-the-prisma-layer-postgresql) to Prisma 2. Then follow up with the upgrade guide to [Prisma 3](https://www.prisma.io/docs/orm/more/upgrade-guides/upgrading-versions/upgrading-to-prisma-3), [Prisma 4](https://www.prisma.io/docs/orm/more/upgrade-guides/upgrading-versions/upgrading-to-prisma-4) and [Prisma 5](https://www.prisma.io/docs/orm/more/upgrade-guides/upgrading-versions/upgrading-to-prisma-5). That way you would have a less painful process.

You may also want to take a look at this previous [discussion](https://github.com/prisma/prisma/discussions/20025) as well.

**Reply by MorayM** [reply]
Hi @ludralph!

Thanks for letting me know about that previous question, I didn't manage to find that when I was searching. I'll have a read through but it looks like it should solve my issue. Looks like I need to re-read the documentation on how the shadow database works!

I originally attempted to do a sequential migration through all the major versions. Unfortunately, versions prior to 4.10.0 require older OpenSSL versions (<3.3) and wouldn't install unless I downgraded my local package, which I was obviously hesitant to do! I figured it was probably easier to just jump straight to v5 and treat it as a brand new installation. My schema is huge and as I'm going to have to rewrite all my database interactions anyway, so better to do it all at once than (potentially) 4 times!
